sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md

```diff
@@ -0,0 +1,436 @@
+    print("Item Name: {}").format(item.name)
+    print("Item Quantity: {}").format(item.quantity)
+    print("Item Price: {}").format(item.total_price)
+print("Subtotal: {}").format(result.subtotal)
+print("Tax: {}").format(result.tax)
+print("Tip: {}").format(result.tip)
+print("Total: {}").format(result.total)
+```
+
+### Layout Models
+
+```python
+class ExtractedLayoutPage:
+    tables: List[ExtractedTable]
+    page_number: int
+
+class ExtractedTable: 
+    List[List[TableCell]]
+    row_count: int
+    column_count: int
+
+class TableCell:
+    text: str
+    column_index: int
+    column_span: int
+    confidence: float
+    is_footer: bool
+    is_header: bool
+    row_index: int
+    row_span: int
+    bounding_box: List[float]
+    raw_field: List[ExtractedLine]
+
+class ExtractedLine:
+    text: str
+    bounding_box: List[float]
+    language: str
+    words: List[ExtractedWord]
+
+class ExtractedWord:
+    text: str
+    bounding_box: List[float]
+    confidence: float
+```
+
+### Layout Sample
+
+```python
+import pandas as pd
+from azure.ai.formrecognizer import FormRecognizerClient
+
+client = FormRecognizerClient(endpoint, credential)
+
+table_image = "https://i.stack.imgur.com/1FyIg.png"
+poller = client.begin_extract_layout(table_image)
+result = poller.result()
+
+dftable = pd.DataFrame(result[0].tables[0]) # page 1, table 1
+print(dftable)
+```
+
+## Custom
+
+The `CustomFormClient` provides all the operations necessary for training a custom model, analyzing a form with a 
+custom model, and managing a user's custom models on their account.
+
+The user can choose to train with or without labels using the methods `begin_labeled_training` or `begin_training`. 
+Both methods take as input a blob SAS uri or valid path to the documents to use for training. Each training method 
+will return a poller object which is used to get the training result.
+
+A custom model can be used to analyze forms using the `begin_extract_form` or `begin_extract_labeled_fields` methods.
+The `model_id` from the training result is passed into the methods, along with the input form to analyze (content-type
+is determined internally). Both methods return a poller object which is used to get the result object.
+
+In order for the user to manage their custom models, a few methods are available to list custom models, delete a model,
+and get a models summary for the account.
+
+### Custom Form Client
+```python
+from azure.ai.formrecognizer import CustomFormClient
+
+client = CustomFormClient(endpoint: str, credential: Union[CognitiveKeyCredential, TokenCredential])
+
+client.begin_labeled_training(
+    source: str, source_prefix_filter: str, include_sub_folders: bool=False
+) -> LROPoller -> LabeledCustomModel
+
+client.begin_training(
+    source: str, source_prefix_filter: str, include_sub_folders: bool=False
+) -> LROPoller -> CustomModel
+
+client.begin_extract_form(form: Any, model_id: str,) -> LROPoller -> List[ExtractedPage]
```

Comment:
Just to clarify - are you saying to include the `readResult` per page in addition to the the `raw_fields` (which contain the rehydrated `readResult` from the `elements` ref)?

Reply:
Yes, that's right. `readResult` has important metadata about the page, like dimensions and orientation.

Reply:
How do you feel about adding these fields as `page_metadata`? Or is there a use case of returning the entire `readResult` on the page?

Reply:
I guess if I went to the trouble of requesting `include_text_details` I would expect to see the full API response.

Scenarios:
1. I am writing an app that compares the full OCR response (all detected words/lines) against the detected keys/values/tables/documents to see if we can improve our annotations.
2. I am building a search engine and I want to capture both structured data (keys/values/tables/documents) and unstructured data (raw words) in my index.
3. I am building a UI that displays keys/values/tables/documents from analyzeResults, but my users need to see the low level words that were detected so they can understand if they have a recognition problem (missing words) or a semantics problem (missing the meaning of the words).

I think it would be a mistake to omit the full `readResult` content from the SDK, and as a user I would be frustrated if there were parts of the Service API that I could not access from the SDK.

Reply:
Building client libraries is a constant question of engineering trade-offs. Many of which are hard. We have had many discussions on if it is worth making the API slightly less user friendly for the 99% use case in order to greatly simplify the 1% use case.

We generally follow the path of "make the easy things easy and make the hard things possible". I don't dispute that the scenarios above are possible. But I would like to get a read as to what percentage of applications do it. And what someone would already have had to understand to use the information. And, at that point, how much of a hinderance is the proposed SDK design going to be.

Reply:
I agree with "make the easy things easy and make the hard things possible" but if we omit part of the API service response (and correct me if I am misunderstanding the proposal) then we are making the hard things _not_ possible using our SDK, right? What is the penalty for including a `page_info` (or whatever) property on the model?

Of course the counter argument to "follow the users as we find them" is that we will end up building a nicer horse carriage instead of a car :)